### PR TITLE
Implement print disasm until optype ##print

### PR DESCRIPTION
**Pull request by eduardvercaemer**
**Checklist**

- [ ] Closing issues: #issue
- [x] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [radare2book](https://github.com/radareorg/radare2book)

**Description**

- Replaces the old `disasm_until_ret` to allow specific optype from `R_ANAL_OP_TYPE_*` members.
- Replaces command `piu` to allow specifying a instruction, like `piu add` to disasm until first `add` opcode.

Example
```
[0x08000040]> pi 11
push rbp
mov rbp, rsp
mov dword [var_ch], 5
mov dword [var_8h], 7
mov edx, dword [var_ch]
mov eax, dword [var_8h]
add eax, edx
mov dword [var_4h], eax
mov eax, dword [var_4h]
pop rbp
ret
[0x08000040]> piuq add
push rbp
mov rbp, rsp
mov dword ptr [rbp - 0xc], 5
mov dword ptr [rbp - 8], 7
mov edx, dword ptr [rbp - 0xc]
mov eax, dword ptr [rbp - 8]
add eax, edx
[0x08000040]> piu pop
0x08000040             push rbp
0x08000041             mov rbp, rsp
0x08000044             mov dword ptr [rbp - 0xc], 5
0x0800004b             mov dword ptr [rbp - 8], 7
0x08000052             mov edx, dword ptr [rbp - 0xc]
0x08000055             mov eax, dword ptr [rbp - 8]
0x08000058             add eax, edx
0x0800005a             mov dword ptr [rbp - 4], eax
0x0800005d             mov eax, dword ptr [rbp - 4]
0x08000060             pop rbp
[0x08000040]> pi?
Usage: pi[bdefrj] [num]
| pia                   print all possible opcodes (byte per byte)
| pib                   print instructions of basic block
| pid                   alias for pdi
| pie                   print offset + esil expression
| pif[?]                print instructions of function
| pij                   print N instructions in JSON
| pir                   like 'pdr' but with 'pI' output
| piu[q] [instruction]  disassemble until instruction is found
| pix  [hexpairs]       alias for pdx and pad
```

### CODE REVIEW

File: libr/core/cmd_print.c [hunk 1]
```diff
@@ -453,7 +453,7 @@ static const char *help_msg_pi[] = {
 	"pif", "[?]", "print instructions of function",
 	"pij", "", "print N instructions in JSON",
 	"pir", "", "like 'pdr' but with 'pI' output",
-	"piu", "[q] [limit]", "disasm until ujmp or ret is found (see pdp)",
+	"piu", "[q] [instruction]", "disassemble until instruction is found",
```
Comment (trufae) [inline libr/core/cmd_print.c:456]: Instruction of given type

**Review by trufae — CHANGES_REQUESTED** [review]
If your pr based on master? This dbg this seems racy somehow i dont see how it should be affected by your pr

File: libr/core/cmd_print.c (hunk 1 above)
Comment (trufae) [inline libr/core/cmd_print.c:456]:
```suggestion
	"piu", "[q] [instruction]", "disassemble until instruction of given optype is found",
```

File: libr/core/cmd_print.c [hunk 2]
```diff
@@ -4188,13 +4188,13 @@ dsmap {
 }
 #endif
 
-static void disasm_until_ret(RCore *core, ut64 addr, char type_print, const char *arg) {
+#define P(x) (core->cons && core->cons->context->pal.x)? core->cons->context->pal.x
+
+static void disasm_until_optype(RCore *core, ut64 addr, char type_print, int optype, int limit) {
+
```
Comment (trufae) [inline libr/core/cmd_print.c:4194]:
```suggestion
```

File: libr/core/cmd_print.c [hunk 3]
```diff
@@ -453,7 +453,7 @@ static const char *help_msg_pi[] = {
 	"pif", "[?]", "print instructions of function",
 	"pij", "", "print N instructions in JSON",
 	"pir", "", "like 'pdr' but with 'pI' output",
-	"piu", "[q] [limit]", "disasm until ujmp or ret is found (see pdp)",
+	"piu", "[q] [instruction]", "disassemble until instruction of given optype is found",
```
Comment (trufae) [inline libr/core/cmd_print.c:456]:
```suggestion
	"piu", "[q] [optype]", "disassemble until instruction of given optype is found (See /atl)",
```

**Review by trufae — APPROVED** [review]

**Comment by trufae** [comment]
Thanks!!The way to list all possible optypes is with /atl, not sure if that would be expected by the user because maybe it should be under `ao` like... adding `aot` command to show the type of the current instruction and then `aotl` to list them all. Adding some tests would be good also for the capstone5 support. Would you like to continue pushing those changes and let me know what do you think about the new commands i suggested

**Comment by eduardvercaemer** [comment]
Thanks for reviewing my PR!
I can work on those commands, maybe also a command to list the instructions that appear in the proximity of the offset ?
I can add the tests however I'm a little confused by the test system, I'll give a read to the `test/` dir and see what i can come up with.

**Pull request merged**